File: ios/SpeakPro/Features/Practice/FollowRead/FollowReadViewModel.swift
import Foundation
import Combine

/// 跟读练习视图模型 —— TTS 参考音播放 + 录音 + ISE 评测
final class FollowReadViewModel: ObservableObject {

    /// 跟读练习的阶段
    enum Phase {
        case ready       // 初始：准备播放参考音
        case listening   // 播放参考音中
        case recording   // 录音中
        case evaluating  // 评测中
        case result      // 显示评测结果
    }

    private var cancellables = Set<AnyCancellable>()

    // MARK: - Published

    @Published var phase: Phase = .ready

    @Published var currentSentence: String = ""
    @Published var currentSentenceIndex: Int = 0
    @Published var totalSentences: Int = 0

    @Published var pronunciationScore: Double = 0
    @Published var intonationScore: Double = 0
    @Published var fluencyScore: Double = 0
    @Published var hasScore = false

    @Published var phonemeErrors: [String] = []
    @Published var isCompleted = false  // 所有句子完成

    @Published var referenceWaveform: [Float] = []
    @Published var studentWaveform: [Float] = []

    // 每句评分历史（用于最终报告）
    struct SentenceScore {
        let sentence: String
        let pronunciation: Double
        let intonation: Double
        let fluency: Double
    }
    @Published var scoreHistory: [SentenceScore] = []

    @Published var isRecording = false
    @Published var isPlayingReference = false
    @Published var isPlayingStudent = false
    @Published var isEvaluating = false
    @Published var errorMessage: String?
    @Published var lastRecordingURL: URL?

    // MARK: - Dependencies

    let audioRecorder = AudioRecorder()
    private let audioPlayer = AudioPlayer()
    private var sentences: [SentenceItem] = []
    private var sessionId: String?

    init() {
        loadSentences()
    }

    // MARK: - 加载句子列表

    func loadSentences() {
        Task { @MainActor in
            do {
                let response: APIResponse<QuestionListResponse> = try await APIClient.shared.get(
                    Endpoints.Questions.list,
                    queryItems: [
                        URLQueryItem(name: "exam_type", value: "IELTS"),
                        URLQueryItem(name: "section", value: "FollowRead"),
                        URLQueryItem(name: "limit", value: "10"),
                    ]
                )

                if let questions = response.data?.items, !questions.isEmpty {
                    sentences = questions.map { SentenceItem(id: $0.id, text: $0.promptText) }
                } else {
                    // 使用默认句子
                    sentences = Self.defaultSentences
                }
                totalSentences = sentences.count
                loadCurrentSentence()
            } catch {
                sentences = Self.defaultSentences
                totalSentences = sentences.count
                loadCurrentSentence()
                print("[FollowReadVM] 句子加载失败: \(error)")
            }
        }
    }

    // MARK: - 导航

    func nextSentence() {
        guard currentSentenceIndex < sentences.count - 1 else { return }
        currentSentenceIndex += 1
        resetScores()
        loadCurrentSentence()
        phase = .ready
    }

    func previousSentence() {
        guard currentSentenceIndex > 0 else { return }
        currentSentenceIndex -= 1
        resetScores()
        loadCurrentSentence()
    }

    private func loadCurrentSentence() {
        guard currentSentenceIndex < sentences.count else { return }
        currentSentence = sentences[currentSentenceIndex].text
    }

    private func resetScores() {
        hasScore = false
        pronunciationScore = 0
        intonationScore = 0
        fluencyScore = 0
        phonemeErrors = []
        studentWaveform = []
        referenceWaveform = []
    }

    // MARK: - 重录

    func retryRecording() {
        hasScore = false
        pronunciationScore = 0
        intonationScore = 0
        fluencyScore = 0
        phonemeErrors = []
        studentWaveform = []
        phase = .ready
    }

    // MARK: - 播放参考音并自动进入录音阶段

    func playReferenceAndTransition() {
        phase = .listening
        errorMessage = nil
        playReference { [weak self] in
            guard let self = self else { return }
            if self.errorMessage != nil {
                DispatchQueue.main.async { self.phase = .ready }
                return
            }
            // 播放完成后进入录音准备阶段（等待用户点击录音按钮）
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                self.phase = .recording
            }
        }
    }

    // MARK: - TTS 参考音播放

    func playReference(completion: (() -> Void)? = nil) {
        if isPlayingReference {
            audioPlayer.stop()
            isPlayingReference = false
            return
        }

        guard !currentSentence.isEmpty else { return }
        isPlayingReference = true

        Task { @MainActor in
            do {
                let response: APIResponse<TTSResponseData> = try await APIClient.shared.post(
                    Endpoints.TTS.synthesize,
                    body: TTSRequestBody(text: currentSentence, speed: 50)
                )

                guard let ttsData = response.data,
                      let audioData = Data(base64Encoded: ttsData.audioB64) else {
                    isPlayingReference = false
                    errorMessage = "TTS 返回数据为空"
                    completion?()
                    return
                }

                // 根据格式选择文件扩展名（Fish Audio 返回 mp3，讯飞返回 pcm）
                let ext = ttsData.format ?? "mp3"
                let tempURL = AudioFileManager.shared.createTempFileURL(extension: ext)
                try audioData.write(to: tempURL)

                // 生成参考波形
                let waveform = WaveformGenerator.generateWaveform(from: tempURL, samplesCount: 60)
                referenceWaveform = waveform

                audioPlayer.play(url: tempURL)
                audioPlayer.$isPlaying
                    .dropFirst()
                    .filter { !$0 }
                    .first()
                    .sink { [weak self] _ in
                        self?.isPlayingReference = false
                        completion?()
                    }
                    .store(in: &cancellables)
            } catch {
                isPlayingReference = false
                errorMessage = "参考音播放失败"
                completion?()
            }
        }
    }

    // MARK: - 播放学生录音

    func playStudentRecording() {
        guard let url = lastRecordingURL else { return }
        if isPlayingStudent {
            audioPlayer.stop()
            isPlayingStudent = false
            return
        }
        isPlayingStudent = true
        audioPlayer.play(url: url)
        audioPlayer.$isPlaying
            .dropFirst()
            .filter { !$0 }
            .first()
            .sink { [weak self] _ in self?.isPlayingStudent = false }
            .store(in: &cancellables)
    }

    // MARK: - 录音

    func startRecording() {
        hasScore = false
        errorMessage = nil
        studentWaveform = []

        Task { @MainActor in
            do {
                try await audioRecorder.requestAndStartRecording()
                isRecording = true
            } catch {
                isRecording = false
                errorMessage = "录音启动失败: \(error.localizedDescription)"
            }
        }
    }

    func stopRecording() {
        guard let audioURL = audioRecorder.stopRecording() else {
            isRecording = false
            phase = .ready
            errorMessage = "录音文件保存失败"
            return
        }
        isRecording = false
        lastRecordingURL = audioURL
        // stopRecording 后波形数据仍保留在 audioRecorder 中
        studentWaveform = audioRecorder.waveformData

        // 检查文件是否有数据
        let fileSize = (try? FileManager.default.attributesOfItem(atPath: audioURL.path)[.size] as? Int) ?? 0
        print("[FollowRead] 停止录音: file=\(audioURL.lastPathComponent), size=\(fileSize) bytes, waveform_points=\(studentWaveform.count)")

        if fileSize < 100 {
            phase = .result
            errorMessage = "录音时间太短（\(fileSize) 字节），请重试"
            return
        }

        phase = .evaluating
        evaluateAudio(fileURL: audioURL)
    }

    // MARK: - 评测

    private func evaluateAudio(fileURL: URL) {
        isEvaluating = true

        Task { @MainActor in
            do {
                guard let audioData = try? Data(contentsOf: fileURL), audioData.count > 100 else {
                    isEvaluating = false
                    phase = .result
                    errorMessage = "录音文件为空（\(((try? Data(contentsOf: fileURL))?.count ?? 0)) 字节），请重试"
                    return
                }

                print("[FollowRead] 录音文件大小: \(audioData.count) 字节, URL: \(fileURL)")

                // 调用发音评测接口
                let evalResp: APIResponse<EvalResponse> = try await APIClient.shared.post(
                    Endpoints.Assessment.evaluate,
                    body: EvalRequestBody(
                        audioB64: audioData.base64EncodedString(),
                        referenceText: currentSentence
                    )
                )

                if let result = evalResp.data {
                    hasScore = true
                    pronunciationScore = result.pronunciationScore?.overall ?? 0
                    intonationScore = result.pronunciationScore?.intonation ?? 0
                    fluencyScore = result.pronunciationScore?.fluency ?? 0

                    // 保存到评分历史
                    scoreHistory.append(SentenceScore(
                        sentence: currentSentence,
                        pronunciation: pronunciationScore,
                        intonation: intonationScore,
                        fluency: fluencyScore
                    ))
                }

                isEvaluating = false
                phase = .result

                // 检查是否全部完成
                if currentSentenceIndex >= totalSentences - 1 {
                    isCompleted = true
                }

                // AI 反馈（可选，失败不影响评分）
                if let sid = sessionId, !sid.isEmpty {
                    do {
                        let feedbackResp: APIResponse<FeedbackResponse> = try await APIClient.shared.post(
                            Endpoints.Assessment.feedback,
                            body: FeedbackRequestBody(
                                sessionId: sid,
                                transcript: currentSentence,
                                referenceText: currentSentence
                            )
                        )
                        if let feedback = feedbackResp.data {
                            phonemeErrors = feedback.corrections ?? []
                        }
                    } catch {
                        print("[FollowRead] AI 反馈获取失败（不影响评分）: \(error)")
                    }
                }
            } catch {
                isEvaluating = false
                phase = .result
                errorMessage = "评测失败: \(error)"
            }
        }
    }

    // MARK: - 默认句子

    private static let defaultSentences: [SentenceItem] = [
        SentenceItem(id: "1", text: "The quick brown fox jumps over the lazy dog."),
        SentenceItem(id: "2", text: "She sells seashells by the seashore."),
        SentenceItem(id: "3", text: "How much wood would a woodchuck chuck?"),
        SentenceItem(id: "4", text: "Peter Piper picked a peck of pickled peppers."),
        SentenceItem(id: "5", text: "I scream, you scream, we all scream for ice cream."),
    ]
}

// MARK: - 内部模型

private struct SentenceItem {
    let id: String
    let text: String
}

private struct QuestionListResponse: Decodable {
    let items: [QuestionItem]?
}

private struct QuestionItem: Decodable {
    let id: String
    let promptText: String
}

private struct TTSRequestBody: Encodable {
    let text: String
    let speed: Int
}

private struct TTSResponseData: Decodable {
    let audioB64: String
    let format: String?
    let sampleRate: Int?
}

private struct EvalRequestBody: Encodable {
    let audioB64: String
    let referenceText: String
}

private struct EvalResponse: Decodable {
    let pronunciationScore: PronScore?
}

private struct PronScore: Decodable {
    let overall: Double?
    let fluency: Double?
    let intonation: Double?
    let integrity: Double?
}

private struct FeedbackRequestBody: Encodable {
    let sessionId: String
    let transcript: String
    let referenceText: String
}

private struct FeedbackResponse: Decodable {
    let corrections: [String]?
    let aiFeedback: String?
}

File: android/app/src/main/java/com/speakpro/features/practice/followread/FollowReadViewModel.kt
package com.speakpro.features.practice.followread

import android.content.Context
import android.media.MediaPlayer
import android.util.Base64
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.speakpro.core.audio.AudioRecorder
import com.speakpro.core.network.ApiClient
import com.speakpro.core.network.ApiClient.goRetrofit
import com.speakpro.core.network.ApiClient.nestRetrofit
import com.speakpro.core.network.Endpoints
import com.speakpro.data.models.ApiResponse
import com.google.gson.annotations.SerializedName
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query
import java.io.File
import javax.inject.Inject

/**
 * 跟读练习阶段
 */
enum class FollowReadPhase {
    READY,       // 准备播放参考音
    LISTENING,   // 播放参考音中
    RECORDING,   // 录音中
    EVALUATING,  // 评测中
    RESULT,      // 显示评测结果
}

/**
 * 单句评分记录
 */
data class SentenceScore(
    val sentence: String,
    val pronunciation: Double,
    val intonation: Double,
    val fluency: Double,
)

/**
 * 跟读练习 ViewModel
 *
 * 流程：加载句子 → 播放参考音 (TTS) → 录音 → 评测 → 显示结果 → 下一句
 */
@HiltViewModel
class FollowReadViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
) : ViewModel() {

    companion object {
        private const val TAG = "FollowReadVM"
    }

    private val audioRecorder = AudioRecorder(context)
    private var lastRecording: File? = null

    // ── 阶段 ──

    private val _phase = MutableStateFlow(FollowReadPhase.READY)
    val phase: StateFlow<FollowReadPhase> = _phase.asStateFlow()

    // ── 句子数据 ──

    private val _currentSentence = MutableStateFlow("")
    val currentSentence: StateFlow<String> = _currentSentence.asStateFlow()

    private val _currentSentenceIndex = MutableStateFlow(0)
    val currentSentenceIndex: StateFlow<Int> = _currentSentenceIndex.asStateFlow()

    private val _totalSentences = MutableStateFlow(0)
    val totalSentences: StateFlow<Int> = _totalSentences.asStateFlow()

    // ── 评分 ──

    private val _pronunciationScore = MutableStateFlow(0.0)
    val pronunciationScore: StateFlow<Double> = _pronunciationScore.asStateFlow()

    private val _intonationScore = MutableStateFlow(0.0)
    val intonationScore: StateFlow<Double> = _intonationScore.asStateFlow()

    private val _fluencyScore = MutableStateFlow(0.0)
    val fluencyScore: StateFlow<Double> = _fluencyScore.asStateFlow()

    private val _hasScore = MutableStateFlow(false)
    val hasScore: StateFlow<Boolean> = _hasScore.asStateFlow()

    private val _phonemeErrors = MutableStateFlow<List<String>>(emptyList())
    val phonemeErrors: StateFlow<List<String>> = _phonemeErrors.asStateFlow()

    private val _isCompleted = MutableStateFlow(false)
    val isCompleted: StateFlow<Boolean> = _isCompleted.asStateFlow()

    // ── 评分历史 ──

    private val _scoreHistory = MutableStateFlow<List<SentenceScore>>(emptyList())
    val scoreHistory: StateFlow<List<SentenceScore>> = _scoreHistory.asStateFlow()

    // ── 播放/录音状态 ──

    private val _isRecording = MutableStateFlow(false)
    val isRecording: StateFlow<Boolean> = _isRecording.asStateFlow()

    private val _isPlayingReference = MutableStateFlow(false)
    val isPlayingReference: StateFlow<Boolean> = _isPlayingReference.asStateFlow()

    private val _isPlayingStudent = MutableStateFlow(false)
    val isPlayingStudent: StateFlow<Boolean> = _isPlayingStudent.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    // ── 波形 ──

    private val _referenceWaveform = MutableStateFlow<List<Float>>(emptyList())
    val referenceWaveform: StateFlow<List<Float>> = _referenceWaveform.asStateFlow()

    val studentWaveform: StateFlow<List<Float>> = audioRecorder.waveformData
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    // ── 内部数据 ──

    private var sentences: List<SentenceItem> = emptyList()
    private var mediaPlayer: MediaPlayer? = null

    init {
        loadSentences()
    }

    // ── 加载句子 ──

    private fun loadSentences() {
        viewModelScope.launch {
            try {
                val api = nestRetrofit.create(FollowReadApi::class.java)
                val response = api.getQuestions(examType = "IELTS", section = "FollowRead", limit = 10)
                val items = response.data?.items
                if (!items.isNullOrEmpty()) {
                    sentences = items.map { SentenceItem(it.id, it.promptText) }
                } else {
                    sentences = defaultSentences
                }
            } catch (e: Exception) {
                Log.w(TAG, "句子加载失败，使用默认句子", e)
                sentences = defaultSentences
            }
            _totalSentences.value = sentences.size
            loadCurrentSentence()
        }
    }

    private fun loadCurrentSentence() {
        val idx = _currentSentenceIndex.value
        if (idx < sentences.size) {
            _currentSentence.value = sentences[idx].text
        }
    }

    // ── 导航 ──

    fun nextSentence() {
        if (_currentSentenceIndex.value >= sentences.size - 1) return
        _currentSentenceIndex.value += 1
        resetScores()
        loadCurrentSentence()
        _phase.value = FollowReadPhase.READY
    }

    fun retryRecording() {
        resetScores()
        _phase.value = FollowReadPhase.READY
    }

    private fun resetScores() {
        _hasScore.value = false
        _pronunciationScore.value = 0.0
        _intonationScore.value = 0.0
        _fluencyScore.value = 0.0
        _phonemeErrors.value = emptyList()
        _referenceWaveform.value = emptyList()
        lastRecording?.delete()
        lastRecording = null
    }

    // ── TTS 参考音播放 + 自动过渡 ──

    fun playReferenceAndTransition() {
        _phase.value = FollowReadPhase.LISTENING
        _errorMessage.value = null
        playReference {
            if (_errorMessage.value != null) {
                _phase.value = FollowReadPhase.READY
                return@playReference
            }
            viewModelScope.launch {
                delay(300)
                _phase.value = FollowReadPhase.RECORDING
            }
        }
    }

    fun playReference(completion: (() -> Unit)? = null) {
        if (_isPlayingReference.value) {
            stopAudio()
            _isPlayingReference.value = false
            return
        }
        if (_currentSentence.value.isEmpty()) return
        _isPlayingReference.value = true

        viewModelScope.launch {
            try {
                val api = goRetrofit.create(FollowReadApi::class.java)
                val resp = api.synthesize(TtsRequestBody(_currentSentence.value, 50))
                val ttsData = resp.data

                if (ttsData == null || ttsData.audioB64.isBlank()) {
                    _isPlayingReference.value = false
                    _errorMessage.value = "TTS 返回数据为空"
                    completion?.invoke()
                    return@launch
                }

                val audioBytes = Base64.decode(ttsData.audioB64, Base64.DEFAULT)
                // 生成参考波形（模拟）
                _referenceWaveform.value = List(60) { (Math.random() * 0.8f + 0.2f).toFloat() }

                playAudioBytes(audioBytes) {
                    _isPlayingReference.value = false
                    completion?.invoke()
                }
            } catch (e: Exception) {
                _isPlayingReference.value = false
                _errorMessage.value = "参考音播放失败"
                completion?.invoke()
            }
        }
    }

    // ── 播放学生录音 ──

    fun playStudentRecording() {
        if (_isPlayingStudent.value) {
            stopAudio()
            _isPlayingStudent.value = false
            return
        }
        val wav = lastRecording
        if (wav == null || !wav.exists()) {
            _errorMessage.value = "暂无可回放的录音"
            return
        }
        _isPlayingStudent.value = true
        viewModelScope.launch(Dispatchers.Main) {
            try {
                mediaPlayer?.release()
                mediaPlayer = MediaPlayer().apply {
                    setDataSource(wav.absolutePath)
                    prepare()
                    setOnCompletionListener { _isPlayingStudent.value = false }
                    start()
                }
            } catch (e: Exception) {
                Log.e(TAG, "录音回放失败", e)
                _isPlayingStudent.value = false
                _errorMessage.value = "录音回放失败"
            }
        }
    }

    // ── 录音 ──

    fun startRecording() {
        _hasScore.value = false
        _errorMessage.value = null
        try {
            audioRecorder.startRecording()
            _isRecording.value = true
        } catch (e: SecurityException) {
            _errorMessage.value = "未授予麦克风权限，请在系统设置中开启"
        } catch (e: Exception) {
            Log.e(TAG, "启动录音失败", e)
            _errorMessage.value = "启动录音失败: ${e.localizedMessage ?: "未知错误"}"
        }
    }

    fun stopRecording() {
        if (!_isRecording.value) return
        _isRecording.value = false
        val wav = audioRecorder.stopRecording()
        lastRecording = wav
        if (wav == null) {
            _errorMessage.value = "未录到有效音频"
            _phase.value = FollowReadPhase.RESULT
            return
        }
        _phase.value = FollowReadPhase.EVALUATING
        evaluateAudio(wav)
    }

    // ── 评测 ──

    private fun evaluateAudio(wav: File) {
        viewModelScope.launch {
            try {
                val audioB64 = withContext(Dispatchers.IO) {
                    Base64.encodeToString(wav.readBytes(), Base64.NO_WRAP)
                }
                val api = goRetrofit.create(FollowReadApi::class.java)
                val resp = api.evaluate(
                    EvalRequestBody(
                        audioB64 = audioB64,
                        referenceText = _currentSentence.value,
                    ),
                )
                val result = resp.data
                if (result != null) {
                    _hasScore.value = true
                    _pronunciationScore.value = result.pronunciationScore?.overall ?: 0.0
                    _intonationScore.value = result.pronunciationScore?.intonation ?: 0.0
                    _fluencyScore.value = result.pronunciationScore?.fluency ?: 0.0

                    _scoreHistory.value = _scoreHistory.value + SentenceScore(
                        sentence = _currentSentence.value,
                        pronunciation = _pronunciationScore.value,
                        intonation = _intonationScore.value,
                        fluency = _fluencyScore.value,
                    )
                }

                _phase.value = FollowReadPhase.RESULT

                if (_currentSentenceIndex.value >= _totalSentences.value - 1) {
                    _isCompleted.value = true
                }
            } catch (e: Exception) {
                _errorMessage.value = "评测失败: ${e.message}"
                _phase.value = FollowReadPhase.RESULT
            }
        }
    }

    // ── 音频辅助 ──

    private fun playAudioBytes(data: ByteArray, onComplete: () -> Unit) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val tmpFile = File.createTempFile("tts_", ".wav")
                tmpFile.writeBytes(data)
                viewModelScope.launch(Dispatchers.Main) {
                    mediaPlayer?.release()
                    mediaPlayer = MediaPlayer().apply {
                        setDataSource(tmpFile.absolutePath)
                        prepare()
                        setOnCompletionListener {
                            onComplete()
                            tmpFile.delete()
                        }
                        start()
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "音频播放失败", e)
                viewModelScope.launch(Dispatchers.Main) { onComplete() }
            }
        }
    }

    private fun stopAudio() {
        mediaPlayer?.stop()
        mediaPlayer?.release()
        mediaPlayer = null
    }

    override fun onCleared() {
        super.onCleared()
        mediaPlayer?.release()
        audioRecorder.release()
        lastRecording?.delete()
    }

    // ── 默认句子 ──

    private val defaultSentences = listOf(
        SentenceItem("1", "The quick brown fox jumps over the lazy dog."),
        SentenceItem("2", "She sells seashells by the seashore."),
        SentenceItem("3", "How much wood would a woodchuck chuck?"),
        SentenceItem("4", "Peter Piper picked a peck of pickled peppers."),
        SentenceItem("5", "I scream, you scream, we all scream for ice cream."),
    )
}

// ── 内部模型 ──

private data class SentenceItem(val id: String, val text: String)

// ── Retrofit API ──

private data class QListResponse(val items: List<QItem>? = null)
private data class QItem(val id: String, @SerializedName("prompt_text") val promptText: String)

private data class TtsRequestBody(val text: String, val speed: Int)
private data class TtsResponseData(
    val audioB64: String = "",
    val format: String? = null,
)

private data class EvalRequestBody(
    val audioB64: String,
    val referenceText: String,
)

private data class EvalResponse(
    // 服务端返回 camelCase（见 assessment.go:145），早期 snake_case 的 @SerializedName 导致解析为 null → 全 0 分
    @SerializedName(value = "pronunciationScore", alternate = ["pronunciation_score"])
    val pronunciationScore: PronScoreResp? = null,
)

private data class PronScoreResp(
    val overall: Double? = null,
    val fluency: Double? = null,
    val intonation: Double? = null,
    val integrity: Double? = null,
)

private interface FollowReadApi {
    @GET(Endpoints.Questions.LIST)
    suspend fun getQuestions(
        @Query("exam_type") examType: String,
        @Query("section") section: String,
        @Query("limit") limit: Int,
    ): ApiResponse<QListResponse>

    @POST(Endpoints.Tts.SYNTHESIZE)
    suspend fun synthesize(@Body body: TtsRequestBody): ApiResponse<TtsResponseData>

    @POST(Endpoints.Assessment.EVALUATE)
    suspend fun evaluate(@Body body: EvalRequestBody): ApiResponse<EvalResponse>
}
